src/main/java/org/entando/kubernetes/service/digitalexchange/job/EntandoBundleInstallService.java

```diff
@@ -311,6 +324,83 @@ private CompletableFuture<Void> submitInstallAsync(EntandoBundleJobEntity parent
+                InstallPlan latestInstallPlan = objectMapper.readValue(latestInstallPlanString, InstallPlan.class);
+                InstallPlan diff = calculateDiffInstallPlan(installPlan, latestInstallPlan);
+
+                Queue<Installable> bundleInstallableComponentsDiff = getBundleInstallableComponents(latestBundleReader, conflictStrategy, diff);
+
+
+                Queue<EntandoBundleComponentJobEntity> componentJobQueueDiff = getBundleComponentJobEntityQueue(parentJob, bundleInstallableComponentsDiff);
+
+                List<ComponentUsage> componentUsageList = getComponentUsageList(componentJobQueueDiff);
+
+                JobScheduler schedulerDiff = new JobScheduler();
+                schedulerDiff.queuePrimaryComponents(componentJobQueueDiff);
+                String bundleRootFolder = BundleUtilities.composeBundleResourceRootFolter(latestBundleReader);
+
+                return uninstallDiff(bundleRootFolder, schedulerDiff, componentUsageList);
+
+            } else {
+                String warnings = "The uninstallation of orphaned components is not necessary since the bundle is not installed on the system";
+                log.debug(warnings);
+                return warnings;
+            }
+        } catch (Exception e) {
+            String error = "The uninstallation of orphaned components failed with error " + e;
+            log.error(error);
+            return error;
+        }
+
+    }
+
+    private List<ComponentUsage> getComponentUsageList(Queue<EntandoBundleComponentJobEntity> componentJobQueueDiff) {
+        List<EntandoBundleComponentJobEntity> componentJobListDiff = new ArrayList<>(componentJobQueueDiff);
+        return usageService.getComponentsUsageDetails(componentJobListDiff).stream()
+                .filter(componentUsage -> {
+                    if (!componentUsage.isExist() || componentUsage.getHasExternal()) {
+                        log.debug(String.format(
+                                "Component '%s' is not found or contains external references so it can't be uninstalled",
+                                componentUsage.getCode()));
+                        return false;
+                    }
+                    return true;
+                }).collect(Collectors.toList());
+    }
+
+    private static Queue<EntandoBundleComponentJobEntity> getBundleComponentJobEntityQueue(EntandoBundleJobEntity parentJob,
+                                                                                           Queue<Installable> bundleInstallableComponentsDiff) {
+        return bundleInstallableComponentsDiff.stream()
+                .filter(i -> {
+                    if (i.getRepresentation() instanceof WidgetDescriptor) {
+                        WidgetDescriptor wd = (WidgetDescriptor) i.getRepresentation();
+                        return TYPE_WIDGET_APPBUILDER.equals(wd.getType());
+                    }
+                    return false;
```

Comment:
Please consider extracting this function to a method with a comprehensive name just to ease the reading process. Might it also be worth adding a sort of `isTypeAppBuilder` within the WidgetDescriptor itself?